### PR TITLE
Trunk-5411:Add doseLimitUnits field to Drug

### PR DESCRIPTION
**Pull request by gitcliff**
link:https://issues.openmrs.org/browse/TRUNK-5411?src=confmacro
Trunk-5411:Add doseLimitUnits field to Drug

### CODE REVIEW

**Comment by dkayiwa** [comment]
You would need to also add a hibernate mapping file in addition to a liquibase changeset.

**Comment by gitcliff** [comment]
thanks @dkayiwa,, going to work on it

**Comment by gitcliff** [comment]
@dkayiwa  i have added the changes you requested in the hibernate mappings and liquibase changeSet

File: pom.xml [hunk 1]
```diff
@@ -531,7 +531,7 @@
 				<plugin>
 					<groupId>org.apache.maven.plugins</groupId>
 					<artifactId>maven-surefire-plugin</artifactId>
-					<version>2.18.1</version>
+					<version>3.0.0-M3</version>
```
Comment (dkayiwa) [inline pom.xml]: Is this change required as part of the ticket?

File: api/src/main/resources/liquibase-update-to-latest.xml [hunk 2]
```diff
@@ -20,5 +20,18 @@
...
-
+<changeSet id="Trunk-5411:Add doseLimitUnits field to Drug" author="gitacliff">
+		<preConditions onFail="MARK_RAN">
+			<not>
+				<columnExists tableName="drug" columnName="dose_Limit_Units"/>				
+			</not>
+		</preConditions>
+		<comment>Adding column doseLimitUnits in table drug</comment>
+        <addColumn tableName="drug">
+            <column name="dose_Limit_Units" type="varchar(255)">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-latest.xml]: Have you looked at the convention that we use for the rest of the table names?

Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: Have you also looked at the database type we use for concepts columns?

File: pom.xml (hunk 1 above)
Comment (gitcliff) [reply pom.xml]: @dkayiwa initially i was getting test failures in the surefire, failing to execute the maven goal in the plugin untill i decided to upgrade it to the latest version as seen here 
https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-surefire-plugin/3.0.0-M3

File: api/src/main/resources/liquibase-update-to-latest.xml (hunk 2 above)
Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: no @dkayiwa i have looked at the conventions for table names here https://wiki.openmrs.org/display/docs/Conventions and its saying lower case singular form, also when you look through the different liquibases for the subsequent openmrs versions its what is being used

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: As for the database type for concept columns i have opted to use Numeric datatype after having a look at https://wiki.openmrs.org/display/docs/Managing+Concept+Datatypes#ManagingConceptDatatypes-HelpfulNotes

Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: Can you show me just one existing liquibase changeset with that naming convention for column names?

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: <changeSet id="20180808-passwordreset" author="harisu+fanyui">
		<preConditions onFail="MARK_RAN">
			<not>
				<columnExists tableName="users" columnName="email"/>				
			</not>
		</preConditions>
		<comment>Adding column users.email</comment>
        <addColumn tableName="users">
            <column name="email" type="varchar(255)">
            	<constraints unique="true"/>
            </column>
        </addColumn>
    </changeSet>
</changeSet>
	<changeSet id="TRUNK-4791-20180216" author="patrick">
		<preConditions onFail="MARK_RAN">
			<not>
				<columnExists tableName="provider" columnName="speciality_id"/>
			</not>
		</preConditions>
		<comment>Adding speciality_id column to provider table</comment>
		<addColumn tableName="provider">
			<column name="speciality_id" type="int" />
		</addColumn>
		<addForeignKeyConstraint baseTableName="provider"
								 baseColumnNames="speciality_id"
								 constraintName="provider_speciality_id_fk"
								 referencedTableName="concept"
								 referencedColumnNames="concept_id" />
	</changeSet>
@dkayiwa  i have gotten these two different changeSets for adding columns to tables users and provider respectively from liquibase-update-to-2.2.xml in the core

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: <changeSet id="TRUNK-4791-20180216" author="patrick">
		<preConditions onFail="MARK_RAN">
			<not>
				<columnExists tableName="provider" columnName="speciality_id"/>
			</not>
		</preConditions>
		<comment>Adding speciality_id column to provider table</comment>
		<addColumn tableName="provider">
			<column name="speciality_id" type="int" />
		</addColumn>
		<addForeignKeyConstraint baseTableName="provider"
								 baseColumnNames="speciality_id"
								 constraintName="provider_speciality_id_fk"
								 referencedTableName="concept"
								 referencedColumnNames="concept_id" />
	</changeSet>

Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: Are they named the same was as your column?

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: <changeSet id="TRUNK-4505-20180804" author="fruether">
		<preConditions onFail="MARK_RAN">
			<not>
				<columnExists tableName="person" columnName="cause_of_death_non_coded"/>
			</not>
		</preConditions>
		<comment>Adding cause_of_death_non_coded to the table person</comment>
		<addColumn tableName="person">
			<column name="cause_of_death_non_coded" type="varchar(255)" />
		</addColumn>
	</changeSet>
@dkayiwa  this changeSet is from the  liquibase-update-to-2.2.xml in the core

**Comment by gitcliff** [comment]
<changeSet id="TRUNK-4505-20180804" author="fruether">
		<preConditions onFail="MARK_RAN">
			<not>
				<columnExists tableName="person" columnName="cause_of_death_non_coded"/>
			</not>
		</preConditions>
		<comment>Adding cause_of_death_non_coded to the table person</comment>
		<addColumn tableName="person">
			<column name="cause_of_death_non_coded" type="varchar(255)" />
		</addColumn>
	</changeSet>
@dkayiwa  this changeSet is from the  liquibase-update-to-2.2.xml in the core

File: api/src/main/resources/liquibase-update-to-latest.xml (hunk 2 above)
Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: Do you notice the difference between the convention for those names and yours?

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: @dkayiwa i stand to be corrected if am going wrong here

Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: OK. Can you write those columns names here and also include yours?

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: no sir, they are not named in the same way as mine column ..
@dkayiwa

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: **the column names :**
<column name="fulfiller_comment" type="varchar(1024)" />
<column name="cause_of_death_non_coded" type="varchar(255)" />
<column name="speciality_id" type="int" />
**mine is:**

<column name="dose_Limit_Units" type="Numeric">
@dkayiwa these are the different column names in comparison with mine

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: <column name="speciality_id" type="int" />
<column name="cause_of_death_non_coded" type="varchar(255)" />
<column name="dose_Limit_Units" type="Numeric">

**Comment by gitcliff** [comment]
<column name="speciality_id" type="int" />
<column name="cause_of_death_non_coded" type="varchar(255)" />
<column name="dose_Limit_Units" type="Numeric">

File: api/src/main/resources/liquibase-update-to-latest.xml (hunk 2 above)
Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: @dkayiwa  i have been trying to send the column names but they are not displaying here ..
Am seeing the difference in the changeset id and the datatype of the column

File: api/src/main/resources/liquibase-update-to-latest.xml [hunk 3]
```diff
@@ -20,5 +20,18 @@
...
+			</not>
+		</preConditions>
+		<comment>Adding column doseLimitUnit in table drug</comment>
+        <addColumn tableName="drug">
+            <column name="dose_Limit_Unit" type="int">
+            	<constraints nullable="false"/>
+            </column>
+        </addColumn>
+    </changeSet>
+
```
Comment (gitcliff) [inline api/src/main/resources/liquibase-update-to-latest.xml]: @dkayiwa are these changes enough to answer your column names naming conventions?

Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: Just list a few column names here together with yours.

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: <column name="cause_of_death_non_coded" type="varchar(255)" />
<column name="speciality_id" type="int" />

  <column name="dose_Limit_Unit" type="int">

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: ooh thanks @dkayiwa,,  i have seen the difference here because of using capital letters

Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: @dkayiwa i have updated the changeSet column name

File: api/src/main/java/org/openmrs/Drug.java [hunk 4]
```diff
@@ -276,4 +278,18 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
 			getDrugReferenceMaps().add(drugReferenceMap);
 		}
 	}
+	
+	/** Returns the doseLimitUnits of a particular drug */
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Can you do proper method javadoc?

File: api/src/main/java/org/openmrs/Drug.java [hunk 5]
```diff
@@ -276,4 +278,18 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
...
+	
+	public Concept getDoseLimitUnits() {
+		return doseLimitUnits;
+	}
+	
+	/**
+	 * @param doseLimitUnits is the doseLimitUnits to be stored in the drug
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Can you do proper method javadoc?

File: api/src/main/resources/liquibase-update-to-2.3.xml [hunk 6]
```diff
@@ -32,5 +32,4 @@
 		<comment>Renaming allerygy.comment to allergy.comments</comment>
     	<renameColumn tableName="allergy" oldColumnName="comment" newColumnName="comments" columnDataType="varchar(1024)"/>
     </changeSet>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.3.xml]: Can you discard these formatting changes?

File: api/src/main/resources/liquibase-schema-only.xml [hunk 7]
```diff
@@ -3436,7 +3436,7 @@
 	<changeSet id="20120529-2231" author="mvorobey">
         <addForeignKeyConstraint constraintName="privilege_which_can_edit_encounter_type" baseTableName="encounter_type" baseColumnNames="edit_privilege" referencedTableName="privilege" referencedColumnNames="privilege"/>
 	</changeSet>
-
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-schema-only.xml]: Can you discard these formatting changes?

File: pom.xml (hunk 1 above)
Comment (dkayiwa) [reply pom.xml]: This change is not related to the ticket and hence should not be in this pull request.

File: api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml [hunk 8]
```diff
@@ -46,11 +46,14 @@
 			column="date_retired" not-null="false" length="19" />
 		<property name="retireReason" type="java.lang.String"
 			column="retire_reason" length="255" />
+		
 
 		<many-to-one name="concept" class="org.openmrs.Concept"
 			not-null="true" column="concept_id" />
 		<many-to-one name="dosageForm" class="org.openmrs.Concept"
 			not-null="false" column="dosage_form" />
+		<many-to-one name="doseLimitUnits" class="org.openmrs.Concept"
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml]: Have you actually run these changes and confirmed that they work?

File: api/src/main/resources/liquibase-update-to-latest.xml [hunk 9]
```diff
@@ -20,5 +20,18 @@
...
     <include file="liquibase-update-to-2.2.xml"/>
     <include file="liquibase-update-to-2.3.xml"/>
-
+<changeSet id="Trunk-5411:Add dose limit unit field to Drug" author="gitacliff">
+		<preConditions onFail="MARK_RAN">
+			<not>
+				<columnExists tableName="drug" columnName="dose_limit_unit"/>				
+			</not>
+		</preConditions>
+		<comment>Adding column dose limit unit in table drug</comment>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-latest.xml]: Shouldn't you also add a foreign key constraint?

File: api/src/main/resources/liquibase-update-to-latest.xml [hunk 10]
```diff
@@ -20,5 +20,18 @@
...
+        <addColumn tableName="drug">
+            <column name="dose_limit_unit" type="int">
+            	<constraints nullable="false"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-latest.xml]: Are you sure that this column should not allow nulls?

File: api/src/main/resources/liquibase-update-to-latest.xml (hunk 9 above)
Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: yes @dkayiwa  am giong to add it

File: api/src/main/resources/liquibase-update-to-latest.xml (hunk 10 above)
Comment (gitcliff) [reply api/src/main/resources/liquibase-update-to-latest.xml]: @dkayiwa my thinking on this is that since this column is going to will represent the units of the existing fields of maximumDailyDose and minimumDailyDose of a given drug and a do believe every drug must these fields catered for which necessitates the doselimitunit not to all nulls

File: api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml (hunk 8 above)
Comment (gitcliff) [reply api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml]: am going to run them

Comment (gitcliff) [reply api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml]: @dkayiwa i have run the changes and the work as seen here https://pastebin.com/ubevYdsX

File: api/src/main/java/org/openmrs/Drug.java [hunk 11]
```diff
@@ -276,4 +278,20 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
 			getDrugReferenceMaps().add(drugReferenceMap);
 		}
 	}
+	
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Can you put a method description?

File: api/src/main/java/org/openmrs/Drug.java [hunk 12]
```diff
@@ -276,4 +278,20 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
...
+	 * @return Returns the doseLimitUnits
+	 * @since 2.3
+	 */
+	public Concept getDoseLimitUnits() {
+		return doseLimitUnits;
+	}
+	
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Can you put a method description?

File: api/src/main/resources/liquibase-update-to-latest.xml [hunk 13]
```diff
@@ -20,5 +20,23 @@
     <include file="liquibase-update-to-2.1.xml"/>
     <include file="liquibase-update-to-2.2.xml"/>
     <include file="liquibase-update-to-2.3.xml"/>
+    <changeSet id="Trunk-5411:Add dose limit unit field to Drug" author="gitacliff">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-latest.xml]: Can you make your changeset id be of the same format like the others?

Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-latest.xml]: The changeset should instead be in the liquibase-update-to-2.3.xml file

**Comment by dkayiwa** [comment]
Can you look into the merge conflicts?

File: api/src/main/resources/liquibase-update-to-2.3.xml [hunk 14]
```diff
@@ -32,5 +32,23 @@
...
+     <changeSet id="20190625-Trunk-5411" author="gitacliff">
+		<preConditions onFail="MARK_RAN">
+			<not>
+				<columnExists tableName="drug" columnName="dose_limit_unit"/>				
+			</not>
+		</preConditions>
+		<comment>Adding column dose limit unit in table drug</comment>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.3.xml]: Can you look at how descriptions look like for existing changesets?

File: api/src/main/resources/liquibase-update-to-2.3.xml [hunk 15]
```diff
@@ -32,5 +32,23 @@
 		<comment>Renaming allerygy.comment to allergy.comments</comment>
     	<renameColumn tableName="allergy" oldColumnName="comment" newColumnName="comments" columnDataType="varchar(1024)"/>
     </changeSet>
+     <changeSet id="20190625-Trunk-5411" author="gitacliff">
+		<preConditions onFail="MARK_RAN">
+			<not>
+				<columnExists tableName="drug" columnName="dose_limit_unit"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.3.xml]: Is it doe limit unit or dose limit units?

**Comment by gitcliff** [comment]
@dkayiwa i have corrected the changes that you requested for

**Comment by dkayiwa** [comment]
The ticket status shows that you are still doing more work on this.

File: api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml [hunk 16]
```diff
@@ -51,6 +51,8 @@
...
+		 not-null="false" column="dose_Limit_Unit" />
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/api/db/hibernate/Drug.hbm.xml]: Do you have any column named like the above?

File: api/src/main/resources/liquibase-update-to-2.3.xml [hunk 17]
```diff
@@ -49,7 +49,7 @@
 		</createTable>
 	</changeSet>
 
-    <changeSet id="TRUNK-5413" author="odorajonathan">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.3.xml]: Why did you make the above changes?

File: api/src/main/java/org/openmrs/Drug.java [hunk 18]
```diff
@@ -276,4 +278,26 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
 			getDrugReferenceMaps().add(drugReferenceMap);
 		}
 	}
+	
+	/**
+	 * Returns the doseLimitUnits for this drug
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Gets the ....

File: api/src/main/resources/liquibase-update-to-2.3.xml [hunk 19]
```diff
@@ -32,8 +32,8 @@
 		<comment>Renaming allerygy.comment to allergy.comments</comment>
     	<renameColumn tableName="allergy" oldColumnName="comment" newColumnName="comments" columnDataType="varchar(1024)"/>
     </changeSet>
-	
-	<changeSet id="TRUNK-5494" author="Wyclif">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.3.xml]: Why did you change these lines?

**Comment by dkayiwa** [comment]
Did you see the travis build failure?

File: api/src/main/resources/liquibase-update-to-2.3.xml (hunk 17 above)
Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-2.3.xml]: Can you discard these changes because they have nothing to do with your ticket?

File: api/src/main/resources/liquibase-update-to-2.3.xml (hunk 19 above)
Comment (dkayiwa) [reply api/src/main/resources/liquibase-update-to-2.3.xml]: Can you discard these changes because they have nothing to do with your ticket?

**Comment by gitcliff** [comment]
@dkayiwa  yes i have seen it , am getting  build success with "mvn clean install" and when i make changes in the travis.yml file to use openjdk8 instead of oraclejdk8 i get successful travis CI, the changes in the travis .yml seem to be  correcting the travis failure yet the file is not part of the ticket...
So guidance on this

**Comment by dkayiwa** [comment]
The travis.yml changes are necessary. But not the others as per my inline comments.

**Comment by gitcliff** [comment]
@dkayiwa thank you for the clarification

File: api/src/main/java/org/openmrs/Drug.java [hunk 20]
```diff
@@ -276,4 +278,24 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
 			getDrugReferenceMaps().add(drugReferenceMap);
 		}
 	}
+	
+	/**
+	 * Gets the doseLimitUnits for this drug
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Shouldn't this match the description that is in the ticket?

File: api/src/main/java/org/openmrs/Drug.java [hunk 21]
```diff
@@ -276,4 +278,24 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
...
+	 * 
+	 * @param Returns the doseLimitUnits.
+	 * @since 2.3.0
+	 */
+	public Concept getDoseLimitUnits() {
+		return doseLimitUnits;
+	}
+	
+	/**
+	 * Sets the doseLimitUnits to this drug
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Shouldn't this match the description that is in the ticket?

**Comment by gitcliff** [comment]
@dkayiwa i have matched the description to that which is in the ticket,,hope its okay to be merged

**Comment by dkayiwa** [comment]
Did you commit? I do not see any changes.

**Comment by gitcliff** [comment]
yes @dkayiwa  i did commit the changes
"Gets the doseLimitUnits field of drug"
"Sets the doseLimitUnits field to drug"

**Comment by dkayiwa** [comment]
Is that how the ticket defines that field?

**Comment by gitcliff** [comment]
@dkayiwa not really,,  i have some changes and requesting for your review.Thanks

File: api/src/main/java/org/openmrs/Drug.java [hunk 22]
```diff
@@ -276,4 +278,26 @@ public void addDrugReferenceMap(DrugReferenceMap drugReferenceMap) {
 			getDrugReferenceMaps().add(drugReferenceMap);
 		}
 	}
+	
+	/**
+	 * Gets doseLimitUnits field to Drug, which will represent the units of the existing fields of
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Drug.java]: Given that this is already the Dug class, what value does Drug add to the decription?

**Comment by gitcliff** [comment]
@dkayiwa  i think the value that Drug adds to the description is doseLimitsUnits

**Comment by dkayiwa** [comment]
Can you read your description again?

**Comment by gitcliff** [comment]
@dkayiwa  i have read the description and made changes again,kindly review

**Comment by dkayiwa** [comment]
What you have done is not put the description that is in the ticket. Take a step back and ask yourself that, If you were the user of this API, what kind of description would you like to see to understand what this field is used for.

**Comment by gitcliff** [comment]
@dkayiwa  is this new description that i have edited okay?,,,,kindly review
